astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
I made some timings for a few tables with different number of rows and columns. Obviously the speedup is better for tables with many columns. c9bd47c corresponds to the first few changes, without caching, and 8fc380a to the last change (caching). Reference is c83e24c.

1R x 720C
ref: 1.07s
c9bd47c: 639ms
8fc380a: 345ms

1R x 133C
ref: 173ms
c9bd47c: 149ms
8fc380a: 141ms

586R x 40C
ref: 21.8ms
c9bd47c: 19.2ms
8fc380a: 18.5ms

854R x 120C
ref: 79.2ms
c9bd47c: 61.2ms
8fc380a: 53.6ms

So, having a mean for faster access of column by name would indeed do the job. I think `names` and `formats` are the two attributes that really benefit from the caching. Should I restrict this cache to these two attributes, or do you see another mean to do it ?